internal/store/vpa.go

```diff
@@ -0,0 +1,421 @@
+					v := c.Target
+					if mem, ok := v[v1.ResourceMemory]; ok {
+						ms = append(ms, &metric.Metric{
+							LabelKeys:   []string{"container_name"},
+							LabelValues: []string{c.ContainerName},
+							Value:       float64(mem.Value()),
+						})
+					}
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_vpa_container_status_recommendation_uncapped_target_cpu_cores",
+			Type: metric.Gauge,
+			Help: "Target CPU cores the VPA recommends for the container ignoring bounds.",
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				ms := []*metric.Metric{}
+				if a.Status.Recommendation == nil || a.Status.Recommendation.ContainerRecommendations == nil {
+					return &metric.Family{
+						Metrics: ms,
+					}
+				}
+				for _, c := range a.Status.Recommendation.ContainerRecommendations {
+					v := c.UncappedTarget
+					if cpu, ok := v[v1.ResourceCPU]; ok {
+						ms = append(ms, &metric.Metric{
+							LabelKeys:   []string{"container_name"},
+							LabelValues: []string{c.ContainerName},
+							Value:       float64(cpu.MilliValue()) / 1000,
+						})
+					}
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+		{
+			Name: "kube_vpa_container_status_recommendation_uncapped_target_memory_bytes",
+			Type: metric.Gauge,
+			Help: "Target memory bytes the VPA recommends for the container ignoring bounds.",
+			GenerateFunc: wrapVPAFunc(func(a *autoscaling.VerticalPodAutoscaler) *metric.Family {
+				ms := []*metric.Metric{}
+				if a.Status.Recommendation == nil || a.Status.Recommendation.ContainerRecommendations == nil {
+					return &metric.Family{
+						Metrics: ms,
+					}
+				}
+				for _, c := range a.Status.Recommendation.ContainerRecommendations {
+					v := c.UncappedTarget
+					if mem, ok := v[v1.ResourceMemory]; ok {
+						ms = append(ms, &metric.Metric{
+							LabelKeys:   []string{"container_name"},
+							LabelValues: []string{c.ContainerName},
+							Value:       float64(mem.Value()),
+						})
+					}
+				}
+				return &metric.Family{
+					Metrics: ms,
+				}
+			}),
+		},
+	}
+)
+
+func wrapVPAFunc(f func(*autoscaling.VerticalPodAutoscaler) *metric.Family) func(interface{}) *metric.Family {
+	return func(obj interface{}) *metric.Family {
+		vpa := obj.(*autoscaling.VerticalPodAutoscaler)
+
+		metricFamily := f(vpa)
+		targetRef := fmt.Sprintf("%s/%s/%s", vpa.Spec.TargetRef.APIVersion, vpa.Spec.TargetRef.Kind, vpa.Spec.TargetRef.Name)
+
+		for _, m := range metricFamily.Metrics {
+			m.LabelKeys = append(descVerticalPodAutoscalerLabelsDefaultLabels, m.LabelKeys...)
+			m.LabelValues = append([]string{vpa.Namespace, vpa.Name, targetRef}, m.LabelValues...)
+		}
+
+		return metricFamily
+	}
+}
+
+func createVPAListWatchFunc(kubeCfg *rest.Config) func(kubeClient clientset.Interface, ns string) cache.ListerWatcher {
+	vpaClient, err := vpaclientset.NewForConfig(kubeCfg)
```

Comment:
this creation should be on the same level as we create the kubeclient, so we can actually properly handle the resulting error